Make updates for GLAMOD release 8.0 -> ready to merge!

ludwiglierhammer opened the pull request:
Speed up the code 🚀:

- [ ] eliminate looping through numpy arrays -> open new Issue https://github.com/glamod/marine_qc/issues/40
- [x] update point selection in climatologies
- [x] update docstrings

ludwiglierhammer committed:
implement parallel processing in Climatology.get_values

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
remove unused imports, rework conds

ludwiglierhammer committed:
resolve merge conflict

ludwiglierhammer committed:
get rid of looping through numpy arrays

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
Merge branch 'main' into release_8

ludwiglierhammer committed:
fiy tests

ludwiglierhammer committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

ludwiglierhammer commented:
@jjk-code-otter: please have a look at the functions I've already changed.

ludwiglierhammer commented:
@jjk-code-otter: feel free to keep eliminating some loops and speeding up the code in this PR.

jjk-code-otter committed:
added finder functions and test for latitude

jjk-code-otter committed:
added test for longitude quick index function

jjk-code-otter committed:
passing test for longitude quick index function

jjk-code-otter committed:
300 times speed up extracting climatology values

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
bug in range checks for get value fixed

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

jjk-code-otter committed:
authors

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
Small bug in IQUAM track check

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
vectorize the find repeated values function

jjk-code-otter committed:
Vectorize the rounded value check

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
fixing bug with missing data in rounded value check

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

jjk-code-otter committed:
Fixing bug in the get value input validation

jjk-code-otter committed:
adding array based versions of some differencers

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
Vectorised versions of calc speed course distance time

jjk-code-otter committed:
Vectorised versions of calc speed course distance time

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
adding course bewteen points calculation from pyproj

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
vector version of lat lon from course distance

jjk-code-otter committed:
trifles

jjk-code-otter committed:
vector version of forward discrepancy

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
backward discrepancy vectored

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer commented:
> I added a different approach for extracting climatological values. The method `get_value_fast` is 300 times faster than `get_value` at extracting 10,000 random lat, lon, date values. It uses the lat and lon coordinates and the number of time points to work out the indices directly into the xarray data array.
> 
> See [6b83d83](https://github.com/glamod/marine_qc/commit/6b83d8331412e9fa61e7fba6ba8158395bac6e8e)

Thanks for this great work 🎉. This is much more than I could have imagined before my holidays 👍 
Reading the climatologies and extracting individual points is one of the biggest challenges using the processing scripts. 

> 
> Also, the original `get_value` function is returning `nan` values for a spatially complete climatology and I'm not sure why (though I haven't looked into it).

Me neither... But that should not be the case under any circumstances. Could you please give me an example of a spatially complete climatology?

ludwiglierhammer commented on AUTHORS.rst:
```diff
@@ -12,3 +12,8 @@ Co-Developers
 -------------
 
 * John J Kennedy <jjk8932@googlemail.com> `@jjk-code-otter <https://github.com/jjk-code-otter>`_
+
+Previous contributors
+---------------------
+
+* Chris Atkinson <chris.atkinson@metoffice.gov.uk>
```
```suggestion
```

ludwiglierhammer replied:
I'll add this in https://github.com/glamod/marine_qc/pull/23.

ludwiglierhammer committed:
add author information in https://github.com/glamod/marine_qc/pull/23

jjk-code-otter committed:
vectorise calculate_midpoint and intermediate_point

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
add pyproj to dependencies

ludwiglierhammer committed:
add more get_value test for other external climatologies

ludwiglierhammer committed:
remove old Climatology

ludwiglierhammer committed:
add more get_value_fast test using other external climatologies

ludwiglierhammer committed:
pre-commit hooks

ludwiglierhammer committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

ludwiglierhammer commented:
I added some tests using other external climatologies (https://github.com/glamod/marine_qc/pull/24/commits/84ef3a77b3e2a67e7f3a4fd5624267c48e772313 and https://github.com/glamod/marine_qc/pull/24/commits/cc54d97745a587ae17612e40472b99b96ae612de). I got the the expected result from the old `Climatology` module, which I deleted in the testing suite. Can we delete it in general or do we need it in any other module, test, etc? Our [daily sst test file](https://github.com/glamod/cdm-testdata/blob/main/metoffice_qc/external_files/SST_daily_climatology_january.nc) has only 31 time steps. This is the way I load it:

```
    return Climatology.open_netcdf_file(
        external_clim["SST"]["mean"],
        "sst",
        valid_ntime=31,
    )
```

In `external_clim.Climatology.get_t_index` we need some adjustments for the case ntime is not `1`, `73` or `365`, I think.

jjk-code-otter committed:
direction continuity vectorised

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
speed continuity

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
check distance from estimate

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
do track check vectorised

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer commented on marine_qc/qc_sequential_reports.py:
```diff
@@ -28,6 +29,249 @@
+    return time_difference.astype(float)
+
+
+def angular_distance_array(lat1, lon1, lat2, lon2):
+    """Return distances in kilometres between points on the globe"""
+    radians_per_degree = np.pi / 180.0
+
+    # convert degrees to radians
+    lat1 = lat1 * radians_per_degree
+    lon1 = lon1 * radians_per_degree
+    lat2 = lat2 * radians_per_degree
+    lon2 = lon2 * radians_per_degree
+
+    delta_lambda = abs(lon1 - lon2)
+    bit1 = np.cos(lat2) * np.sin(delta_lambda)
+    bit1 = bit1 * bit1
+    bit2 = np.cos(lat1) * np.sin(lat2) - np.sin(lat1) * np.cos(lat2) * np.cos(
+        delta_lambda
+    )
+    bit2 = bit2 * bit2
+
+    top_bit = bit1 + bit2
+    top_bit = np.sqrt(top_bit)
+
+    bottom_bit = np.sin(lat1) * np.sin(lat2) + np.cos(lat1) * np.cos(lat2) * np.cos(
+        delta_lambda
+    )
+
+    return np.arctan2(top_bit, bottom_bit)
+
+
+def sphere_distance_array(lat1, lon1, lat2, lon2):
+    """Return distances in kilometres between points on the globe"""
+    earths_radius = 6371.0088
+    return angular_distance_array(lat1, lon1, lat2, lon2) * earths_radius
```
```suggestion
geod = Geod(ellps="WGS84")
def geod_inv(lon1, lat1, lon2, lat2):
    """Returns forward azimuth, back azimuth, and distance  using the ellipsoidal model."""
    fwd_az, back_az, dist = geod.inv(lon1, lat1, lon2, lat2)
    return fwd_az, back_az, dist

def angular_distance_array(lat1, lon1, lat2, lon2):
    """Returns angular distance in radians between two points using the ellipsoidal model."""
    return geod_inv(lon1, lat1, lon2, lat2)[2] / geod.a  # geod.a: ellipsoid equatorial radius
    
def sphere_distance_array(lat1, lon1, lat2, lon2):
    """Returns geodesic distance in kilometers using the ellipsoidal model."""
    return geod_inv(lon1, lat1, lon2, lat2)[2] / 1000.  # convert to kilometers
```

ludwiglierhammer replied:
We can use pyproj as we did in `course_between_points_array`. 
Problem: ``test_sphere_distance_array`` fails since the earth is not a perfect sphere but`pyproj` does work with an ellipsiod.

ludwiglierhammer replied:
We also could postpone this in another issue. Maybe, that could be to much for this PR.

jjk-code-otter replied:
The level of precision we get from a non-spherical earth is more precision than we need and it makes the routines (and the routines that depend on it) very hard to test.

Presumably pyproj can do a spherical earth. That might be a better option in this case if it's possible.

jjk-code-otter replied:
https://pyproj4.github.io/pyproj/stable/advanced_examples.html#caching-pyproj-objects

ludwiglierhammer commented on marine_qc/external_clim.py:
```diff
@@ -270,6 +287,103 @@ def convert_units_to(self, target_units, source_units=None) -> None:
+        day_arr = np.where(np.isnan(day_arr), -1, day_arr).astype(int)
+
+        ml = get_month_lengths(2004)
+        month_lengths = np.array([ml[m - 1] if 1 <= m <= 12 else 0 for m in month_arr])
```
This is some more vectorization.

ludwiglierhammer commented on marine_qc/qc_sequential_reports.py:
```diff
@@ -28,6 +29,249 @@
+
+def intermediate_point_array(lat1, lon1, lat2, lon2, f):
+
+    earths_radius = 6371.0088
+    radians_per_degree = np.pi / 180.0
+
+    f[f > 1.0] = np.nan
+
+    d = angular_distance_array(lat1, lon1, lat2, lon2)
+
+    lat1 = lat1 * radians_per_degree
+    lon1 = lon1 * radians_per_degree
+    lat2 = lat2 * radians_per_degree
+    lon2 = lon2 * radians_per_degree
+
+    a = np.sin((1 - f) * d) / np.sin(d)
+    b = np.sin(f * d) / np.sin(d)
+    x = a * np.cos(lat1) * np.cos(lon1) + b * np.cos(lat2) * np.cos(lon2)
+    y = a * np.cos(lat1) * np.sin(lon1) + b * np.cos(lat2) * np.sin(lon2)
+    z = a * np.sin(lat1) + b * np.sin(lat2)
+    lat = np.arctan2(z, np.sqrt(x * x + y * y)) / radians_per_degree
+    lon = np.arctan2(y, x) / radians_per_degree
+
+    lat[d == 0.0] = lat1[d == 0.0] / radians_per_degree
+    lon[d == 0.0] = lon1[d == 0.0] / radians_per_degree
+
+    return lat, lon
```
```suggestion
    # Clamp f to [0, 1]
    f = np.clip(f, 0, 1)

    fwd_az, back_az, dist = geod.inv(lon1, lat1, lon2, lat2)
    distance_at_f = dist * f

    lon_f, lat_f, _ = geod.fwd(lon1, lat1, fwd_az, distance_at_f)
    return lat_f, lon_f
```

ludwiglierhammer commented on marine_qc/qc_sequential_reports.py:
```diff
@@ -28,6 +29,249 @@
+
+
+def lat_lon_from_course_and_distance_array(lat1, lon1, tc, d):
+    radians_per_degree = np.pi / 180.0
+    earths_radius = 6371.0088
+
+    lat1 = lat1 * radians_per_degree
+    lon1 = lon1 * radians_per_degree
+    tcr = tc * radians_per_degree
+
+    dr = d / earths_radius
+
+    lat = np.arcsin(np.sin(lat1) * np.cos(dr) + np.cos(lat1) * np.sin(dr) * np.cos(tcr))
+    dlon = np.arctan2(
+        np.sin(tcr) * np.sin(dr) * np.cos(lat1), np.cos(dr) - np.sin(lat1) * np.sin(lat)
+    )
+    lon = np.mod(lon1 + dlon + np.pi, 2.0 * np.pi) - np.pi
+
+    lat = lat / radians_per_degree
+    lon = lon / radians_per_degree
+
+    return lat, lon
```
```suggestion
    dist_m = d * 1000
    
    lon2, lat2, back_az = geod.fwd(lon1, lat1, tc, dist_m)
    return lat2, lon2
```

ludwiglierhammer commented on marine_qc/external_clim.py:
```diff
@@ -270,6 +287,103 @@ def convert_units_to(self, target_units, source_units=None) -> None:
+        n_points = len(month)
+        t_index = np.zeros(n_points)
+
+        if ntime == 1:
+            return t_index
+        elif ntime == 73:
+            return which_pentad_array(month, day)
+        elif ntime == 365:
+            return day_in_year_array(month=month, day=day)
```
It is just for testing, yes. For the moment, I'll adapt your approach. Maybe we can think about making it more flexible in the future if needed.

jjk-code-otter replied:
We could have a special case for 31 for the test, but a more general approach would be better. I couldn't think how to easily do it. The lat and lon axes are regular so it's easy, but having a function that can deal with daily, pentad, months, 31 days, weeks or whatever a climatology might be, is much harder.

jjk-code-otter committed:
Update marine_qc/external_clim.py

Co-authored-by: Ludwig Lierhammer <Ludwig.Lierhammer@dwd.de>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
workaround: expand January data to full year

ludwiglierhammer committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

ludwiglierhammer committed:
convert ml to numpy array

jjk-code-otter committed:
fix small bug

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

ludwiglierhammer commented:
We worked on the same issue (https://github.com/glamod/marine_qc/pull/24/commits/553bfeb10bc82e7584d9d6a22fcb22b63876c537 and https://github.com/glamod/marine_qc/pull/24/commits/92068a09b2593d73227382328636c14be80b2529) 😬

ludwiglierhammer committed:
vectorize modal_speed

ludwiglierhammer committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

jjk-code-otter commented on marine_qc/qc_individual_reports.py:
```diff
@@ -28,6 +26,11 @@
 from .external_clim import ClimFloatType, inspect_climatology
 from .time_control import convert_date, day_in_year, get_month_lengths
 
+vectorized_day_in_year = np.vectorize(day_in_year)
+vectorized_sunangle = np.vectorize(
+    sunangle, otypes=[float, float, float, float, float, float]
+)
```
Why not? I had a quick look at the sunangle calculation and it could be vectorised one function at a time, but I don't know what the tradeoff is here in time spent refactoring and time gained in running. In principle, we can test the time saved by vectorizing, but I don't think it's necessary to do it for every single function.

ludwiglierhammer committed:
use pyproj for angular and sphere distance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
use pyproj for intermediate_point

ludwiglierhammer committed:
use pyproj

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
set earths radius in auxiliary

ludwiglierhammer committed:
make use a spheric earth instead od ellipsiod

ludwiglierhammer committed:
use np.allclose instead of np.all

ludwiglierhammer committed:
use older lat_lon_from_course_and_distance_array version

ludwiglierhammer committed:
use earths radius in meter

ludwiglierhammer committed:
ml only once to a array

ludwiglierhammer committed:
optionally, set dtype in post_format_return_type

ludwiglierhammer committed:
inspect_climatology with get_value_fast

ludwiglierhammer commented:
Task:
 - [ ] Tommorrow, I'll fix the failing tests. They all depend on external daily SST January file.

ludwiglierhammer committed:
workaround: set ntime to 365 in SST testdata

ludwiglierhammer committed:
get_t_index returns exact indices starting from 0

ludwiglierhammer committed:
mask not avaiable indices

ludwiglierhammer commented:
@jjk-code-otter: could you please have a look if https://github.com/glamod/marine_qc/pull/24/commits/e436983015276f2c0d24788355c43d8295764cc7 is reasonable. I updated `Climatology.get_t_index` to return indices starting with 0.

ludwiglierhammer committed:
use external_clim instead of old Climatology

jjk-code-otter committed:
docstring

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

jjk-code-otter committed:
Removing Climatology.py

ludwiglierhammer commented:
I think, we can now replace `get_value` with `get_value_fast` and remove the older version.

jjk-code-otter commented:
> I think, we can now replace `get_value` with `get_value_fast` and remove the older version.

get_value_fast only works in a specific set of circumstances whereas get_value should work for any xarray dataset (right?). I would suggest keeping both for flexibility even if we only intend to use the fast version.

jjk-code-otter committed:
small addition to get_t_index test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
Moving array versions of track check related functions

jjk-code-otter commented:
> Move some functions:
> 
> * [x]  `time_differences_array`  to `time_control`
> * [x]  `angular_distance_array` to `spherical_geometry`
> * [x]  `sphere_distance_array` to `spherical_geometry`
> * [x]  `intermediate_point_array` to `spherical_geometry`
> * [x]  `course_between_points_array` to `spherical_geometry`
> * [x]  `increment_position_array` to `track_check_utils`

see 7d50536

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer commented:
> While I am moving things, I could also move the following out of `qc_sequential_reports.py` and into `track_check_utils.py`
> 
> * [ ]  calculate_speed_course_distance_time_difference_array
> * [ ]  forward_discrepancy_array
> * [ ]  backward_discrepancy_array
> * [ ]  calculate_midpoint_array
> * [ ]  calculate_course_parameters
> * [ ]  calculate_speed_course_distance_time_difference
> * [ ]  forward_discrepancy
> * [ ]  backward_discrepancy
> * [ ]  calculate_midpoint
> 
> Bascially, everything that isn't actually a QC check.
> 
> @ludwiglierhammer What do you think?

That's a good idea. Where do you move time to? Any current submodels in your mind or do you create a pontand to `track_check_utils`?

ludwiglierhammer committed:
use get_value_fast instead of get_value

ludwiglierhammer commented:
> > I think, we can now replace `get_value` with `get_value_fast` and remove the older version.
> 
> get_value_fast only works in a specific set of circumstances whereas get_value should work for any xarray dataset (right?). I would suggest keeping both for flexibility even if we only intend to use the fast version.

What afre those specific circumstances. At the first glance, `get_value_fast` looks good to me. In https://github.com/glamod/marine_qc/pull/24/commits/ae93796bfd671cb60233dfcd49bd9ee49d6fd696, I replaced `get_value` with `get_value_fast`. All tests are passing except `test_qc_on_db.buddy_check`. Do you have an explanation?  `i=11`: results: `1`; expected: `0`

jjk-code-otter committed:
moving functions out of qx_sequential_reports

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
fix test for buddy check

ludwiglierhammer commented:
> > What afre those specific circumstances. At the first glance, `get_value_fast` looks good to me. In [ae93796](https://github.com/glamod/marine_qc/commit/ae93796bfd671cb60233dfcd49bd9ee49d6fd696), I replaced `get_value` with `get_value_fast`.
> 
> I was thinking that maybe someone might want to use an irregular grid at some point in the future. There are lots of strange ocean model grids out there. I assume xarray can deal with these OK, but our `fast` approach would not work at all.

That is correct but `get_value` is very weak in performance. I thought of implementing a method that select nearest points instead if necessary. Here is a coarse example:

```
def select_nearest(data, time_coords, lat_coords, lon_coords, time_val, lat_val, lon_val):
    t_idx = np.abs(time_coords - time_val).argmin()
    lat_idx = np.abs(lat_coords - lat_val).argmin()
    lon_idx = np.abs(lon_coords - lon_val).argmin()
    return data[t_idx, lat_idx, lon_idx]
```

I think, something like this is more useful than keeping `get_value`. What do you think?

jjk-code-otter commented:
> Do you think, it is reasonable to update the expected test values (see [14141df](https://github.com/glamod/marine_qc/commit/14141df02767e4189df6043a2603a901696558fc)) if the results differ because we need another method? I am sceptical.

Fair question. The test here wasn't carefully constructed, I just copied and pasted the output from the buddy check as it was at that point in time and looked at the arrays to see if it made some kind of sense. The xarray driven `get_value` function had different behaviour to what was previously used in the code (`mds_lon_to_yindex` and similar) so we might expect situations where the different methods give different answers. I think it's OK in this situation to change the test.

ludwiglierhammer commented:
> > I replaced `get_value` with `get_value_fast`. All tests are passing except `test_qc_on_db.buddy_check`. Do you have an explanation? `i=11`: results: `1`; expected: `0`
> 
> One of the observations fell on a grid cell boundary. What I think happened is that the fast and slow versions assign it to different grid cells and that makes a difference in one observation. I changed the test in this case because it was there to catch and highlight changes like this, but I'm not sure we are that fussy about how to do the assignment.

Ok, for me that is fine. Changing expected test result makes a programmer shudder slightly 😁. But I think, I should trust more in your expertise in quality control. So, it is fine 👍

ludwiglierhammer commented:
> > Do you think, it is reasonable to update the expected test values (see [14141df](https://github.com/glamod/marine_qc/commit/14141df02767e4189df6043a2603a901696558fc)) if the results differ because we need another method? I am sceptical.
> 
> Fair question. The test here wasn't carefully constructed, I just copied and pasted the output from the buddy check as it was at that point in time and looked at the arrays to see if it made some kind of sense. The xarray driven `get_value` function had different behaviour to what was previously used in the code (`mds_lon_to_yindex` and similar) so we might expect situations where the different methods give different answers. I think it's OK in this situation to change the test.

I see. Other buddy check tests still work.

jjk-code-otter committed:
adding docstrings to array versions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter commented:
> I tested the marine processing QC script (just the part with the individual observations). With the older version it took more than 20 minutes, and now it takes less than 1 minute 🎉 🚀.
> 
> That is pretty nice 😀.
> 
> Thanks a lot!

That is pretty nice! Woo hoo! 

Which are the slow bits now?

ludwiglierhammer commented:
> > I tested the marine processing QC script (just the part with the individual observations). With the older version it took more than 20 minutes, and now it takes less than 1 minute 🎉 🚀.
> > That is pretty nice 😀.
> > Thanks a lot!
> 
> That is pretty nice! Woo hoo!
> 
> Which are the slow bits now?

The individual observations are running pretty good.
I haven't tested sequential or grouped reports yet.
Tomorrow I'll test the sequential ones.

ludwiglierhammer commented:
> > Do we need those duplicate functions like `do_track_check` and `do_track_check_array`? Can we remove the ones without `_array`?
> 
> Only if they're not used any more.

I'll clean this up and update the tests.

ludwiglierhammer committed:
optionally, format multiple reuslts arrays

ludwiglierhammer committed:
impoer inspect_arrays from auxiliary

ludwiglierhammer committed:
replace * functions with *_array functions

ludwiglierhammer commented:
@jjk-code-otter:  we can update all docstrings and fix the pre-commit issues when the tests are running again.

ludwiglierhammer committed:
vectorize find_saturated_runs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
solve runtime warning

jjk-code-otter committed:
fixed spherical geometry tests

jjk-code-otter committed:
fixed track check tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
Adding minute-long changes to time diff test from failing track check examples

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer commented on tests/test_spherical_geometry.py:
```diff
@@ -149,7 +149,7 @@ def test_heading_east():
     """Heading due west should be course of 90degrees"""
     heading = sg.course_between_points(0.0, 0.0, 0.0, -23.2)
-    assert pytest.approx(heading, 0.0000001) == 270.0
+    assert pytest.approx(heading, 0.0000001) == 270.0 or pytest.approx(heading, 0.0000001) == -90.0
```
@jjk-code-otter: Thank for fixing the tests. Do you think, it is helpful/reasonable that `course between_points` return values between `0` and `360` to avid `-90` for instance? Or is the difference between `270` and `-90` important?

jjk-code-otter replied:
See #34

ludwiglierhammer committed:
optionally, ignore indexes in buddy checks

ludwiglierhammer committed:
add try statement

ludwiglierhammer committed:
add ZeroDivisionError

jjk-code-otter commented:
> @jjk-code-otter, in [8c9236a](https://github.com/glamod/marine_qc/commit/8c9236a7b1fabc49e26e74e19df0933828fdfbf5), I added a try statement since I got this error:
> 
> ValueError: Upper limit is less than x: r_hi = 8.0, x = 40.419189453125.
> 
> Have you any comments/ideas?
> 
> Edit: I added `ZeroDivisionError`

The bayesian buddy check is intended to be used in combination with a climatology check that has limits at +-8C. If an anomaly is outside this range, the code will produced nonsensical values, so there is code to ensure that doesn't happen. 

In the original code, basic individual report QC was done first and then only those observations that passed a pre-determined set of checks were passed to the track check and buddy check. In that case, no observations with an anomaly >8C reached the bayesian buddy check.

ludwiglierhammer commented:
> > @jjk-code-otter, in [8c9236a](https://github.com/glamod/marine_qc/commit/8c9236a7b1fabc49e26e74e19df0933828fdfbf5), I added a try statement since I got this error:
> > ValueError: Upper limit is less than x: r_hi = 8.0, x = 40.419189453125.
> > Have you any comments/ideas?
> > Edit: I added `ZeroDivisionError`
> 
> The bayesian buddy check is intended to be used in combination with a climatology check that has limits at +-8C. If an anomaly is outside this range, the code will produced nonsensical values, so there is code to ensure that doesn't happen.
> 
> In the original code, basic individual report QC was done first and then only those observations that passed a pre-determined set of checks were passed to the track check and buddy check. In that case, no observations with an anomaly >8C reached the bayesian buddy check.

Thanks for the explantion. To be on the safe side, I would keep the try statement.

jjk-code-otter committed:
docstrings added fstrings tidied

jjk-code-otter committed:
docstrings

ludwiglierhammer committed:
return untestable if any parameter is invalid

ludwiglierhammer committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

jjk-code-otter committed:
additional invalid parameter test

ludwiglierhammer committed:
add sigma_m test

jjk-code-otter committed:
plotting and changelog

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
docstrings for plots routines

jjk-code-otter committed:
merging

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
Adding docstring

jjk-code-otter committed:
Merge branch 'release_8' of https://github.com/ludwiglierhammer/marine_qc into release_8

jjk-code-otter committed:
debugging plotter

ludwiglierhammer committed:
buddy checks: continue if anomality is invalid

ludwiglierhammer committed:
add marplotlib to requirements

ludwiglierhammer committed:
add labels

ludwiglierhammer committed:
remove unused imports

ludwiglierhammer commented:
@jjk-code-otter: do you think is is reasonable: https://github.com/glamod/marine_qc/pull/24/commits/7f73a6d6d235df70fae4cd3385d1ba8ee7f0e03d

ludwiglierhammer committed:
Merge branch 'main' into release_8

ludwiglierhammer committed:
make suplots

ludwiglierhammer commented:
@jjk-code-otter: Could we implement `ignore_indexes` in:
- `SuperObsGrid().add_multiple_observations`
-  `SuperObsGrid().get_new_buddy_limits`
- `SuperObsGrid().get_buddy_limits_with_parameters`

In our GLAMOD example, we don't need to check C-RAID data.

jjk-code-otter committed:
start vectorise buddy check

jjk-code-otter committed:
merge

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jjk-code-otter committed:
Adding rapid checks

jjk-code-otter committed:
breaking changes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
Adding docstrings

ludwiglierhammer committed:
remove unused import

ludwiglierhammer committed:
convert date to pandas.Timestamp

ludwiglierhammer commented:
@jjk-code-otter: I fixed it in https://github.com/glamod/marine_qc/pull/24/commits/772ddf2d19a34c45eb51de3c438e62dde6f8d7c8.

I had to convert a ``numpy.datetime`` to a ``pd.Timestamp`` since ``date.year`` is not valid for a `numpy.datetime`.

ludwiglierhammer committed:
minot

ludwiglierhammer committed:
update docstrings

ludwiglierhammer committed:
update CHANGELOT

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
Update CHANGES.rst

ludwiglierhammer committed:
return warning instead of error if external climatology is not avaiable

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ludwiglierhammer committed:
test if array is empty

ludwiglierhammer committed:
update CHANGELOG

ludwiglierhammer committed:
fixing 404 errors
